File: aws-ses-unsubscribe-lambda.js
import { SESv2Client, PutSuppressedDestinationCommand } from "@aws-sdk/client-sesv2";

// Initialize the SES v2 Client
// AWS Lambda automatically provides credentials and region through the execution environment
const ses = new SESv2Client({ region: process.env.AWS_REGION || "us-east-1" });

export const handler = async (event) => {
    console.log("Received SNS event:", JSON.stringify(event, null, 2));

    for (const record of event.Records) {
        try {
            // Extract the SES message published to the SNS topic
            const snsMessageRaw = record.Sns.Message;
            const snsMessage = JSON.parse(snsMessageRaw);

            // SES Unsubscribe events typically have eventType 'Subscription'
            // when interacting with the {{amazonSESUnsubscribeUrl}} macro
            const eventType = snsMessage.eventType || snsMessage.notificationType;

            if (eventType !== "Subscription" && eventType !== "Unsubscribe") {
                console.log(`Ignoring event type: ${eventType}`);
                continue;
            }

            // Extract the email address(es) that triggered the unsubscribe
            // SES events generally provide the original destination email in the mail.destination array
            let emailsToSuppress = [];

            if (snsMessage.mail && snsMessage.mail.destination) {
                emailsToSuppress = snsMessage.mail.destination;
            } else if (snsMessage.subscription && snsMessage.subscription.contactList) {
                // Fallback or alternative structure if the destination array is formatted differently
                emailsToSuppress = [snsMessage.mail?.destination?.[0] || ""];
            }

            // Filter out any empty/undefined emails
            emailsToSuppress = emailsToSuppress.filter(Boolean);

            for (const email of emailsToSuppress) {
                console.log(`Adding ${email} to SES Account-Level Suppression List...`);

                // Add the email to the Account-level suppression list (PutSuppressedDestination)
                // Note: The SES v2 API only accepts 'BOUNCE' or 'COMPLAINT' as suppression reasons.
                // We use 'COMPLAINT' here as the closest representation of an explicit user opt-out.
                const command = new PutSuppressedDestinationCommand({
                    EmailAddress: email,
                    Reason: "COMPLAINT"
                });

                await ses.send(command);
                console.log(`Successfully suppressed ${email}`);
            }

        } catch (error) {
            console.error("Error processing record:", error);
            // In production, you may want to re-throw the error or send to a DLQ depending on requirements
        }
    }

    // Return success response for Lambda
    return {
        statusCode: 200,
        body: "SNS event processed successfully."
    };
};

File: aws_ses_unsubscribe_lambda.py
import json
import logging
import boto3

logger = logging.getLogger()
logger.setLevel(logging.INFO)

# Initialize SES v2 client
# Boto3 implicitly uses the execution role credentials provided by AWS Lambda
ses_client = boto3.client('sesv2')

def lambda_handler(event, context):
    logger.info(f"Received SNS event: {json.dumps(event)}")
    
    for record in event.get('Records', []):
        try:
            # Parse the SNS message payload
            sns_message_str = record.get('Sns', {}).get('Message')
            if not sns_message_str:
                continue
                
            sns_message = json.loads(sns_message_str)
            
            # Identify the event type.
            # When a user clicks {{amazonSESUnsubscribeUrl}}, SES automatically
            # categorizes it as a 'Subscription' event type.
            event_type = sns_message.get('eventType')
            
            if event_type not in ['Subscription', 'Unsubscribe']:
                logger.info(f"Ignoring non-subscription event type: {event_type}")
                continue
                
            # Extract the email address(es) involved in this unsubscribe event
            # Normally found under 'mail' -> 'destination' in SES notification JSON
            emails_to_suppress = []
            
            if 'mail' in sns_message and 'destination' in sns_message['mail']:
                emails_to_suppress.extend(sns_message['mail']['destination'])
            
            # Clean up potentially empty values and avoid duplicates
            emails_to_suppress = list(set([str(e).strip() for e in emails_to_suppress if e]))
            
            for email in emails_to_suppress:
                logger.info(f"Adding {email} to SES Account-Level Suppression List...")
                
                # Use Amazon SES v2 API to add to the account-level suppression list.
                # The SES v2 API accepts two Reasons: 'BOUNCE' or 'COMPLAINT'.
                # We use 'COMPLAINT' to functionally enforce an absolute block on future sending.
                response = ses_client.put_suppressed_destination(
                    EmailAddress=email,
                    Reason='COMPLAINT'
                )
                logger.info(f"Successfully suppressed {email}: {json.dumps(response)}")
                
        except Exception as e:
            logger.error(f"Error processing record: {str(e)}")
            # Raise the exception so Lambda records a failure and can optionally retry
            raise
            
    return {
        'statusCode': 200,
        'body': json.dumps('Processed SES unsubscribe events successfully.')
    }
